iceoryx2-services/discovery/src/service_discovery/tracker.rs

```diff
@@ -0,0 +1,128 @@
+
+use iceoryx2::{
+    config::Config,
+    prelude::CallbackProgression,
+    service::{service_id::ServiceId, Service, ServiceDetails},
+};
+use std::collections::{HashMap, HashSet};
+
+/// Errors that can occur during service synchronization.
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum SyncError {
+    /// The caller does not have permissions to look up services.
+    InsufficientPermissions,
+
+    /// Failure looking up services present in the iceoryx2 system.
+    ServiceLookupFailure,
+}
+
+/// A tracker for monitoring services of a specific type.
+///
+/// The `Tracker` keeps track of services in the system, allowing for discovery
+/// of new services and detection of services that are no longer available.
+///
+/// # Type Parameters
+///
+/// * `S` - The type of service to track, which must implement the `Service` trait
+pub struct Tracker<S: Service> {
+    services: HashMap<ServiceId, ServiceDetails<S>>,
+}
+
+impl<S: Service> Default for Tracker<S> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<S: Service> Tracker<S> {
+    /// Create a new Monitor instance.
+    pub fn new() -> Self {
+        Self {
+            services: HashMap::new(),
+        }
+    }
+
+    /// Synchronizes the tracker with the current state of services in the system.
+    ///
+    /// This method queries the system for all services of type `S` and updates the tracker's
+    /// internal state. It identifies new services that have appeared since the last sync
+    /// and services that are no longer available.
+    ///
+    /// # Arguments
+    ///
+    /// * `config` - Configuration used to discover services
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing:
+    /// * A vector of service IDs that were newly discovered, details are stored in the tracker and
+    ///   retrievable with `Tracker::get()`
+    /// * A vector of service details for services that are no longer available, these details are
+    ///   no longer stored in the tracker
+    pub fn sync(
+        &mut self,
+        config: &Config,
+    ) -> Result<(Vec<ServiceId>, Vec<ServiceDetails<S>>), SyncError> {
+        let mut discovered_ids = HashSet::<ServiceId>::new();
+        let mut added_ids = Vec::<ServiceId>::new();
+
+        S::list(config, |service| {
+            let id = service.static_details.service_id().clone();
+            discovered_ids.insert(id.clone());
+
+            // Track new services.
+            if !self.services.contains_key(&id) {
+                self.services.insert(id.clone(), service);
+                added_ids.push(id);
+            }
+            CallbackProgression::Continue
+        })
+        .map_err(|e| match e {
+            iceoryx2::service::ServiceListError::InsufficientPermissions => {
+                SyncError::InsufficientPermissions
+            }
+            iceoryx2::service::ServiceListError::InternalError => SyncError::ServiceLookupFailure,
+        })?;
```

Comment:
This could also be refactored to use the `From` trait

Reply:
Squashed into: https://github.com/eclipse-iceoryx/iceoryx2/pull/703/commits/5b00835df9c6492c24b690d1f8c888074a43bab2